pandas/core/internals/managers.py

```diff
@@ -66,7 +68,618 @@
+        # for block in self.blocks:
+        #     if block._verify_integrity and block.shape[1:] != mgr_shape[1:]:
+        #         raise construction_error(tot_items, block.shape[1:], self.axes)
+        # if len(self.items) != tot_items:
+        #     raise AssertionError(
+        #         "Number of manager items must equal union of "
+        #         f"block items\n# manager items: {len(self.items)}, # "
+        #         f"tot_items: {tot_items}"
+        #     )
+
+    def reduce(self: T, func) -> T:
+        # TODO this still fails because `func` assumes to work on 2D arrays
+        assert self.ndim == 2
+
+        res_arrays = []
+        for array in self.arrays:
+            res = func(array)
+            res_arrays.append(np.array([res]))
+
+        index = Index([0])  # placeholder
+        new_mgr = type(self)(res_arrays, [index, self.items])
+        return new_mgr
+
+    def apply(self: T, f, align_keys=None, **kwargs) -> T:
+        """
+        Iterate over the blocks, collect and create a new BlockManager.
+
+        Parameters
+        ----------
+        f : str or callable
+            Name of the Block method to apply.
+
+        Returns
+        -------
+        BlockManager
+        """
+        assert "filter" not in kwargs
+
+        align_keys = align_keys or []
+        result_arrays: List[ExtensionArray] = []
+        # fillna: Series/DataFrame is responsible for making sure value is aligned
+
+        aligned_args = {k: kwargs[k] for k in align_keys}
+
+        if f == "apply":
+            f = kwargs.pop("func")
+
+        for a in self.arrays:
+
+            if aligned_args:
+                # TODO
+                raise NotImplementedError
+
+            if callable(f):
+                applied = f(a, **kwargs)
+            else:
+                applied = getattr(a, f)(**kwargs)
+            result_arrays.append(applied)
+
+        if len(result_arrays) == 0:
+            return self.make_empty(self._axes)
+
+        return type(self)(result_arrays, self._axes)
+
+    def isna(self, func) -> "BlockManager":
+        return self.apply("apply", func=func)
+
+    def where(
+        self, other, cond, align: bool, errors: str, try_cast: bool, axis: int
+    ) -> "ArrayManager":
+        # TODO can be shared
+        if align:
+            align_keys = ["other", "cond"]
+        else:
+            align_keys = ["cond"]
+            other = extract_array(other, extract_numpy=True)
+
+        return self.apply(
+            "where",
+            align_keys=align_keys,
+            other=other,
+            cond=cond,
+            errors=errors,
+            try_cast=try_cast,
+            axis=axis,
+        )
+
+    def replace(self, value, **kwargs) -> "ArrayManager":
+        assert np.ndim(value) == 0, value
+        # TODO "replace" is right now implemented on the blocks, we should move
+        # it to general array algos so it can be reused here
```

Comment:
maybe for the POC could make a `apply_with_block` that wraps in `Block` before doing `apply`?

Reply:
Good idea, will try that tomorrow!